### PR TITLE
move type coercion of agg and agg_udaf to logical phase

### PR DESCRIPTION
**Pull request by liukun4515**
# Which issue does this PR close?

Closes https://github.com/apache/arrow-datafusion/issues/3752
Closes https://github.com/apache/arrow-datafusion/issues/3623

 # Rationale for this change

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

**Comment by liukun4515** [comment]
PTAL @alamb

File: datafusion/physical-expr/src/aggregate/coercion_rule.rs [hunk 1]
```diff
@@ -1,54 +0,0 @@
-// Licensed to the Apache Software Foundation (ASF) under one
```
Comment (alamb) [inline datafusion/physical-expr/src/aggregate/coercion_rule.rs:1]: 👍

File: datafusion/optimizer/src/type_coercion.rs [hunk 2]
```diff
@@ -596,6 +659,123 @@ mod test {
...
+            "MY_AVG",
+            DataType::Float64,
+            Arc::new(DataType::Float64),
+            Volatility::Immutable,
+            Arc::new(|_| Ok(Box::new(AvgAccumulator::try_new(&DataType::Float64)?))),
+            Arc::new(vec![DataType::UInt64, DataType::Float64]),
+        );
+        let udaf = Expr::AggregateUDF {
+            fun: Arc::new(my_avg),
+            args: vec![lit(10i64)],
```
Comment (alamb) [inline datafusion/optimizer/src/type_coercion.rs:675]: 👍

File: datafusion/optimizer/src/type_coercion.rs [hunk 3]
```diff
@@ -596,6 +659,123 @@ mod test {
...
+            distinct: false,
+            filter: None,
+        };
+        let plan =
+            LogicalPlan::Projection(Projection::try_new(vec![agg_expr], empty, None)?);
+        let rule = TypeCoercion::new();
+        let mut config = OptimizerConfig::default();
+        let plan = rule.optimize(&plan, &mut config)?;
+        assert_eq!(
+            "Projection: AVG(Int64(12))\n  EmptyRelation",
```
Comment (alamb) [inline datafusion/optimizer/src/type_coercion.rs:738]: I don't see any coercion happening here. Maybe I am missing something

**Review by alamb — APPROVED** [review]

**Review by andygrove — APPROVED** [review]

File: datafusion/optimizer/src/type_coercion.rs (hunk 3 above)
Comment (liukun4515) [reply datafusion/optimizer/src/type_coercion.rs:738]: Yes, you don't missing anything.

You can take a look `type_coercion::aggregates::coerce_types` function which just check the input data type and don't do any coercion for the function.

**Pull request merged**